Add transitive whitelisting to Santa

nguyen-phillip opened the pull request:
Creates a new queue to listen for compiler-related events in santa-driver (exec, close, and rename), and these are passed on to SNTCompilerController which uses them to determine when it should auto-whitelist compiled binaries.

nguyen-phillip commented:
For testing purposes:
sudo santactl rule --compiler --force --path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld                     
sudo santactl rule --compiler --force --path /usr/local/go/pkg/tool/darwin_amd64/link

tburgin commented on Santa.xcodeproj/project.pbxproj:
```diff
@@ -1051,11 +1057,31 @@
 				LastUpgradeCheck = 0730;
 				TargetAttributes = {
+					0D0016A1192BCD3C005E7FCD = {
+						DevelopmentTeam = 9L79GAXCBC;
```
Remove any DevelopmentTeam additions

nguyen-phillip replied:
Done.

tburgin commented on Source/common/SNTKernelCommon.h:
```diff
@@ -48,7 +48,8 @@ enum SantaDriverMethods {
   QUEUETYPE_DECISION,
-  QUEUETYPE_LOG
+  QUEUETYPE_LOG,
+  QUEUETYPE_COMPILER,
```
As discussed we should be able to get rid of the compiler queue and using the decision queue. This only works if we are able to reduce the number of compiler messages sent over the queue by filtering based on known compiler pids in the kernel.

tburgin commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -14,6 +14,13 @@
+// because that was the first unused integer from sys/kauth.h.  The reason that we don't
+// simply use the KAUTH_VNODE_WRITE_DATA constant as is is because it overlaps with the other
+// KAUTH_FILEOP  constants.
+#define KAUTH_FILEOP_WRITE 8
```
If this is still needed we should give the actions used by KAUTH_SCOPE_FILEOP room to grow. I don't really like the idea of adding our own action, but if we need to make it more than 8 incase fileop adds a new action in the future.

Maybe KAUTH_FILEOP_DELETE + 32 or something

nguyen-phillip replied:
Done.

tburgin commented on Source/santactl/Commands/SNTCommandRule.m:
```diff
@@ -61,15 +62,21 @@ + (NSString *)longHelpText {
 - (void)runWithArguments:(NSArray *)arguments {
   SNTConfigurator *config = [SNTConfigurator configurator];
-  if ([config syncBaseURL] && ![arguments containsObject:@"--check"]) {
+  // TODO: --force flag was added so that we could manually add compiler rules during testing.
```
You can add the --force flag in an `#ifdef DEBUG` block so it can only be used on debug builds.

nguyen-phillip replied:
Done.

nguyen-phillip commented:
PTAL.

Also for testing with Xcode: 
sudo santactl rule --force --compiler --path /usr/bin/codesign
And for rust:
sudo santactl rule --force --whitelist --path ~/.cargo/bin/rustc                                                                                                                                                                                                                                                                                     
sudo santactl rule --force --whitelist --path ~/.rustup/toolchains/stable-x86_64-apple-darwin/bin/rustc

tburgin commented on Source/santad/SNTPolicyProcessor.h:
```diff
@@ -44,11 +44,29 @@
+///        artifacts on disk. Simply pass nil to fileInfo and pass in the desired SHA256s.
+///
+- (nonnull SNTCachedDecision *)decisionForFileInfo:(nullable SNTFileInfo *)fileInfo
+                                fileSHA256:(nullable NSString *)fileSHA256
```
colon align

nguyen-phillip replied:
Done.

nguyen-phillip commented on Source/common/SNTRule.h:
```diff
@@ -49,4 +55,9 @@
+///
+///  Sets timestamp of rule to the current time.
+///
+- (void)refreshTimestamp;
```
Renamed `refreshTimestamp` to `resetTimestamp`.  I made timestamp into a readonly property, but SNTRuleTable needs to be able to set it with results from the rules database.  So this meant adding an extra initializer for SNTRule.

I like having the method to reset the timestamp inside of SNTRule so that the logic for setting the timestamp value is self-contained.  Otherwise we'd have to duplicate the code to set the timestamp in two places, which could lead to errors if we ever decided to change the format.

nguyen-phillip commented on Source/common/SNTRule.m:
```diff
@@ -64,7 +68,10 @@ - (BOOL)isEqual:(id)other {
+  return ([self.shasum isEqual:o.shasum] &&
+          self.state == o.state &&
+          self.type == o.type &&
+          self.timestamp == o.timestamp);
```
Nope, removed.

nguyen-phillip commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -14,6 +14,14 @@
+// FileOpCallback method.  The KAUTH_FILEOP_* constants are defined in
+// sys/kauth.h and run from 1--7.  KAUTH_VNODE_WRITE_DATA is already defined as
+// 4 so it overlaps with the other KAUTH_FILEOP_* constants and can't be used.
+// We define KAUTH_FILEOP_WRITE as something >> 7.
```
Fixed.  Sorry -- bad shorthand for" much greater than"

nguyen-phillip commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -286,6 +297,13 @@ santa_action_t SantaDecisionManager::GetFromCache(uint64_t identifier) {
         decision_cache->remove(identifier);
         return ACTION_UNSET;
       }
+    } else if (result == ACTION_RESPOND_ALLOW_TRANSITIVE) {
```
It's because the the timestamps on the transitive rules only get updated when the rules are retrieved from the rules database.  If a transitive rule ends up in the decision cache and manages to stay in the cache for a long period of time, then its timestamp will never be updated even though the user might be running the binary it applies to fairly often.  Setting an expiration time on the transitive rules forces them to be periodically retrieved from the database, keeping its last access timestamp fresh.

nguyen-phillip commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -427,6 +445,89 @@ void SantaDecisionManager::DecrementListenerInvocations() {
+  struct timespec ts;         // time to sleep between checks
+  SantaDecisionManager *sdm;  // reference to SantaDecisionManager
+} pid_monitor_info;
+
```
Fixed.  (BTW, this struct definition and the following function might should be moved somewhere else since they're not part of the class, I just liked having all the process monitoring code close together).

nguyen-phillip commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -427,6 +445,89 @@ void SantaDecisionManager::DecrementListenerInvocations() {
+// Function to monitor for process termination and then remove the process pid
+// from cache of compiler pids.
+static void pid_monitor(void *param, __unused wait_result_t wait_result) {
+  pid_monitor_info *info = (pid_monitor_info *)param;
```
Done.

nguyen-phillip commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -427,6 +445,89 @@ void SantaDecisionManager::DecrementListenerInvocations() {
+
+  // Kernel logging with IOLog appears broken, so send log message to userspace.
+  // TODO: remove logging.
+  auto message = info->sdm->NewMessage(nullptr);
```
Done.

nguyen-phillip commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -427,6 +445,89 @@ void SantaDecisionManager::DecrementListenerInvocations() {
+  auto info = new pid_monitor_info;
+  info->pid = pid;
+  info->ts = ts_;
+  info->sdm = this;
```
That's a good point. We could have SantaDecisionManager keep track of all the threads it creates and then kill them on exit, but that could be a lot of threads & we'd need to keep track of when treads were finished in order to remove them from the active list.  We maybe could also use a global var to store the compiler pids instead?

nguyen-phillip commented on Source/santa-driver/SantaDecisionManager.h:
```diff
@@ -150,6 +161,14 @@ class SantaDecisionManager : public OSObject {
+    be daily refreshed from the rules database, to keep their last accessed
+    timestamp in the database fresh.
+  */
+  static const uint64_t kMaxAllowTransitiveCacheTimeMilliseconds = 86400000;
```
A write to the vnode would mean that we'd want to get rid of the associated transitive rule anyway most likely.  Reboots might not happen for a long time.  So the only event that would happen often enough is receiving a blacklist rule, but I'm not sure I'm comfortable relying on something that is a somewhat randomly occurring for keeping timestamps up to date.  Exec logs might be a viable alternative.

nguyen-phillip commented on Source/santad/DataLayer/SNTRuleTable.m:
```diff
@@ -101,13 +108,34 @@ - (NSUInteger)certificateRuleCount {
 
+- (NSUInteger)compilerRuleCount {
+  __block NSUInteger count = 0;
+  [self inDatabase:^(FMDatabase *db) {
```
I didn't realize that longForQuery: took varargs -- switched to that.

nguyen-phillip commented on Source/santad/DataLayer/SNTRuleTable.m:
```diff
@@ -101,13 +108,34 @@ - (NSUInteger)certificateRuleCount {
+- (NSUInteger)transitiveRuleCount {
+  __block NSUInteger count = 0;
+  [self inDatabase:^(FMDatabase *db) {
+    count = [db longForQuery:
```
Switched to using varargs.

nguyen-phillip commented on Source/santad/SNTApplication.m:
```diff
@@ -185,6 +190,11 @@ - (void)beginListeningForDecisionRequests {
+          case ACTION_NOTIFY_CLOSE:
+            // Determine if we should add a transitive whitelisting rule for this new file.
+            // Requires that writing process was a compiler and that new file is executable.
+            [self.compilerController checkForNewExecutable:message];
```
Done.

nguyen-phillip commented on Source/santad/SNTCompilerController.m:
```diff
@@ -0,0 +1,65 @@
+
+  // Check if this file is an executable.
+  SNTFileInfo *fi = [[SNTFileInfo alloc] initWithPath:@(target)];
+  if (fi.isExecutable) {
```
Done.

nguyen-phillip commented on Source/santad/SNTCompilerController.m:
```diff
@@ -0,0 +1,65 @@
+    // Check if there is an existing (non-transitive) rule for this file.  We leave existing rules
+    // alone, so that a whitelist or blacklist rule can't be overwritten by a transitive one.
+    SNTRuleTable *ruleTable = [SNTDatabaseController ruleTable];
+    SNTRule *prevRule = [ruleTable ruleForBinarySHA256:fi.SHA1 certificateSHA256:nil];
```
Yikes! Done.

nguyen-phillip commented on Source/santad/SNTCompilerController.m:
```diff
@@ -0,0 +1,65 @@
+                                          customMsg:@""];
+
+    // Add the new rule to the rules database.
+    NSError *err = [[NSError alloc] init];
```
Done.

nguyen-phillip commented:
Fixed all the small stuff, plus added in code for getting compiler rules and transitive whitelisting config from server.  Need to spend some more time thinking about how to deal with the abandoned thread issue and keeping the transitive rule timestamps fresh.

nguyen-phillip commented:
Fixed the transitive rule timestamps so that they get reset for every execution via the event logger.  This means that the ACTION_RESPOND_ALLOW_TRANSITIVE type is not strictly necessary any more and could be removed, unless there is some future reason why we might want it.

Changed the process monitor threads to access a global variable for the compiler pid set, and stopped passing in a reference to SantaDecisionManager, so that they would be safer.  

An alternative to avoid the global variable would be to add a counting semaphore to SantaDecisionManager to keep track of the number of threads spun off.  Each thread would decrement the counter on exit.  When SantaDecisionManager wanted to die, it would set a flag that all threads would read and use to decide to stop, then SantaDecisionManager would wait until the counter decremented to 0.

tburgin commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -14,9 +14,51 @@
+// from cache of compiler pids.
+static void pid_monitor(void *param, __unused wait_result_t wait_result) {
+  pid_t pid = (pid_t)(uintptr_t)param;
+  struct timespec ts = { .tv_sec = 1, .tv_nsec = 0 }; // wait 1 sec between each poll
```
1 sec seems like a long time: a PID could be recycled and another process will then start writing files creating unwanted transitive rules. At the same time `proc_find` is an expensive call for the whole system since it acquires the global proc list lock. So we don't want to be hammering it too often.

Maybe we could use PID + HASH(proc_name or random seed) for the compiler_pid_set_ key. We could then leave the polling at 1 sec, giving us greater confidence that a PID recycle would not result in compiler status being granted to an unworthy process.

We could store the PID in the lower 32 bits and the hash in the upper 32. This way we can extract the PID easily for lookup with proc_find.

nguyen-phillip replied:
My thinking was that it would probably take longer than 1 sec for the PIDs to wrap back around.  I think there are 100,000 of them?  Though I suppose if you had a lot of processes already running, then the number of open PIDs might be more like 90,000.  The problem with adding a hash to the key is that we won't be able to recreate the key when we need to look up the PIDs, because we don't always have access to the proc_name.  I guess we could shorten the time?  I haven't run any tests with different times to see how it might affect performance.

tburgin commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -14,9 +14,51 @@
+  pid_t pid = (pid_t)(uintptr_t)param;
+  struct timespec ts = { .tv_sec = 1, .tv_nsec = 0 }; // wait 1 sec between each poll
+
+  while (true) {
```
If `compiler_pid_set_` goes away we should break the look and kill the thread.

nguyen-phillip replied:
Done.

tburgin commented on Source/santad/SNTExecutionController.m:
```diff
@@ -109,25 +113,46 @@ - (void)validateBinaryWithMessage:(santa_message_t)message {
+      [[SNTConfigurator configurator] transitiveWhitelistingEnabled]) {
+    // If rule indicated that the allowed binary was a compiler and Santa is configured to allow
+    // transitive whitelisting, upgrade the action to ACTION_RESPOND_ALLOW_COMPILER. When sent back
+    // to the kernel, it can use this info to cache the vnode for filtering of future messages.
```
the last sentence is a little unclear

nguyen-phillip replied:
I'm not even sure what I meant here, so I just removed it.

tburgin commented:
Do you mind resolving the conflict in `Source/santad/SNTEventLog.m`?

tburgin commented:
After the outstanding comments are resolved this LGTM. A little cleanup surrounding DEBUG builds will need to happen when hooking the feature into a sync server. Performs well under load.

Are you able to add any more tests?

We should also have @russellhancox hammer on this a bit before merging.

nguyen-phillip commented on Source/common/SNTRule.m:
```diff
@@ -19,17 +19,37 @@ @implementation SNTRule
+                         state:(SNTRuleState)state
+                          type:(SNTRuleType)type
+                     customMsg:(NSString *)customMsg {
+
```
Done.

nguyen-phillip commented on Source/common/SNTRule.m:
```diff
@@ -64,7 +86,9 @@ - (BOOL)isEqual:(id)other {
-  return ([self.shasum isEqual:o.shasum] && self.state == o.state && self.type == o.type);
+  return ([self.shasum isEqual:o.shasum] &&
+          self.state == o.state &&
+          self.type == o.type);
```
Done.

nguyen-phillip commented on Source/common/SNTRule.m:
```diff
@@ -73,12 +97,19 @@ - (NSUInteger)hash {
   result = prime * result + [self.shasum hash];
   result = prime * result + self.state;
   result = prime * result + self.type;
+  result = prime * result + self.timestamp;
```
Oops, forgot to remove this.

nguyen-phillip commented on Source/common/SNTRule.m:
```diff
@@ -73,12 +97,19 @@ - (NSUInteger)hash {
-  return [NSString stringWithFormat:@"SNTRule: SHA-256: %@, State: %ld, Type: %ld",
-                                    self.shasum, self.state, self.type];
+  return [NSString stringWithFormat:@"SNTRule: SHA-256: %@, State: %ld, Type: %ld, Timestamp: %lu",
+          self.shasum, self.state, self.type, (unsigned long)self.timestamp];
```
Done.

nguyen-phillip commented on Source/common/SNTRule.m:
```diff
@@ -73,12 +97,19 @@ - (NSUInteger)hash {
+# pragma mark Last-access Timestamp
+
+- (void)resetTimestamp {
+  _timestamp = (NSUInteger)[[NSDate date] timeIntervalSinceReferenceDate];
```
Done.

nguyen-phillip commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -14,9 +14,51 @@
+// Function to monitor for process termination and then remove the process pid
+// from cache of compiler pids.
+static void pid_monitor(void *param, __unused wait_result_t wait_result) {
+  pid_t pid = (pid_t)(uintptr_t)param;
```
Done.

russellhancox commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -14,9 +14,52 @@
+// This keeps track of all pids associated with compiler processes.  It is defined as a global
+// variable so that the pid monitor threads can access it without needing to reference
+// our instance of SantaDecisionManager.
+static SantaCache<bool> *compiler_pid_set_ = new SantaCache<bool>(500, 5);
```
As discussed:
* make this a property of SDM
* add a counter to SDM (like listener_invocations_) to track the number of threads that are outstanding
* atomically inc/dec that counter before making and terminating the threads
* during SDM dealloc (or ClientDisconnect, maybe?) make this cache pointer NULL to make the threads begin terminating and then sleep-loop until the number of outstanding threads drops to 0.
* use a struct to pass the SDM reference and the pid to the threads when they start

nguyen-phillip replied:
Done. Added pid_monitor_thread_count_ to track the number of threads.  Added PidMonitorThreadsShouldExit() function for the threads to check, since cache pointer is private.  Currently calling StopPidMonitorThreads() from SDM's free() method.  Alternately I think maybe this could be called from SantaDriver::stop() similar to StopListener.  Not sure which is better.

russellhancox commented on Source/santad/SNTCompilerController.h:
```diff
@@ -0,0 +1,25 @@
+#import <Foundation/Foundation.h>
+#import "SNTKernelCommon.h"
+
+@class SNTDriverManager;
```
Leftover?

nguyen-phillip replied:
Removed.

russellhancox commented on Source/santad/SNTExecutionController.m:
```diff
@@ -109,25 +113,46 @@ - (void)validateBinaryWithMessage:(santa_message_t)message {
+  } else if (cd.decision == SNTEventStateAllowTransitive) {
+    // We also upgrade transitive rule decisions.  These are filtered by SNTPolicyProcessor before
+    // they get here so that we only receive them if transitive whitelisting is enabled.
+    // TODO(nguyenphillip): may not need to distinguish transitive allows in the kernel anymore.
```
Do we?

nguyen-phillip replied:
Nope.  Got rid of ACTION_RESPOND_ALLOW_TRANSITIVE and just store SNTEventStateAllowTransitive decisions as ACTION_RESPOND_ALLOW in SantaDecisionManager's decision cache.

russellhancox commented on Source/santad/SNTExecutionController.m:
```diff
@@ -109,25 +113,46 @@ - (void)validateBinaryWithMessage:(santa_message_t)message {
+  if (cd.decision == SNTEventStateAllowCompiler &&
+      [[SNTConfigurator configurator] transitiveWhitelistingEnabled]) {
+    // If rule indicated that the allowed binary was a compiler and Santa is configured to allow
+    // transitive whitelisting, upgrade the action to ACTION_RESPOND_ALLOW_COMPILER.
```
Could SNTPolicyProcessor not make this determination?

nguyen-phillip replied:
Moved the check for transitiveWhitelistingEnabled to SNTPolicyProcessor so that if it's not enabled, the decision passed to SNTExecutionController from a SNTRuleStateWhitelistCompiler rule will be downgraded to SNTEventStateAllow.

nguyen-phillip commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -14,9 +14,52 @@
+    proc_t proc = proc_find(pid);
+    if (!proc) break;
+    proc_rele(proc);
+    msleep(param, NULL, 0, "", &ts);
```
Done.  Added kPidMonitorSleepTimeMilliseconds constant as a private member, and PidMonitorSleepTimeMilliseconds() as a public method to return it.

nguyen-phillip commented:
PTAL

russellhancox commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -208,6 +219,83 @@ kern_return_t SantaDecisionManager::StopListener() {
+static void pid_monitor(void *param, __unused wait_result_t wait_result) {
+  pid_monitor_info *info = (pid_monitor_info *)param;
+  uint32_t sleep_time = info->sdm->PidMonitorSleepTimeMilliseconds();
+  if (info->sdm) {
```
To be on the safe side I'd check info is not nullptr first too

nguyen-phillip replied:
Done.

russellhancox commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -208,6 +219,83 @@ kern_return_t SantaDecisionManager::StopListener() {
+
+  // Keep track of how many times we've slept waiting for the pid monitor
+  // threads to exit.  If we sleep for more than 5 seconds, give up on waiting.
+  int wait_count = 0;
```
This worries me - if the thread didn't decrement the thread count because it hasn't been scheduled there's a possibility it'll be scheduled after SDM is freed and then attempt to access it which will almost certainly panic.

nguyen-phillip replied:
Took out the counter.  I was worried that SDM might get stuck waiting forever if something went wrong with the thread count, but I suppose that is probably preferable to a kernel panic.  Added code to increase the sleep time to 5 sec if the wait loop is still going after 10 sec, in case this does get stuck.

russellhancox commented on Source/santa-driver/SantaDecisionManager.h:
```diff
@@ -290,13 +328,18 @@ class SantaDecisionManager : public OSObject {
   struct timespec ts_;
+
+  // When set to true, Santa will check all ancestors of a process to determine
+  // if it is a compiler.
```
There doesn't appear to be a way to make this true - is it just for testing or are we going to add a driver method to enable it?

nguyen-phillip replied:
Eventually there should be a way to change it along with other parameters that are currently defined as constants.  Added a TODO for this and moved this to live with the other constants.

russellhancox commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -208,6 +219,83 @@ kern_return_t SantaDecisionManager::StopListener() {
+// Arguments that are passed to pid_monitor thread.
+typedef struct {
+  pid_t pid;                  // process to monitor
+  SantaDecisionManager *sdm;  // reference to SantaDecisionManager
```
Can you add a TODO somewhere appropriate to split this out of SDM at some point? I'd like to make SDM smaller and I think this can almost entirely live on its own.

nguyen-phillip replied:
Done.

nguyen-phillip commented on Source/santa-driver/SantaDecisionManager.cc:
```diff
@@ -208,6 +219,83 @@ kern_return_t SantaDecisionManager::StopListener() {
+// termination and then remove the process pid from cache of compiler pids.
+static void pid_monitor(void *param, __unused wait_result_t wait_result) {
+  pid_monitor_info *info = (pid_monitor_info *)param;
+  uint32_t sleep_time = info->sdm->PidMonitorSleepTimeMilliseconds();
```
Done.

nguyen-phillip commented on Source/common/SNTConfigurator.h:
```diff
@@ -159,6 +159,15 @@ extern NSString *const kMobileConfigFilePath;
+///  whitelist any executables that they produce.  If NO, SNTRuleStateWhitelistCompiler rules are
+///  interpreted as if they were simply SNTRuleStateWhitelist rules.  Defaults to NO.
+///
+@property BOOL transitiveWhitelistingEnabled;
```
Done. added key to both with a DEBUG guard around the syncServerKeys entry to be removed when the sync server supports it.

nguyen-phillip commented:
Also, here's an example of what the logs look like now for executing "go run hello.go" where a temporary allow rule is being stored in the kernel while SNTCompilerController figures out the permanent rule.

[2018-02-01T21:14:15.339Z] I santad: action=EXEC|decision=ALLOW|reason=COMPILER|sha256=9f16595bf07331ddb52968c67a478b3a4bc2a3c234a8d1b12341ff0d80b92e57|pid=32312|ppid=32\
306|uid=484912|user=nguyenphillip|gid=5000|group=eng|mode=L|path=/usr/local/go/pkg/tool/darwin_amd64/link|args=/usr/local/go/pkg/tool/darwin_amd64/link -o /var/folders/9\
f/0snyvwy57519np346scs_fh800fsjh/T/go-build271032812/command-line-arguments/_obj/exe/hello -L /var/folders/9f/0snyvwy57519np346scs_fh800fsjh/T/go-build271032812 -s -w -e\
xtld=clang -buildmode=exe -buildid=508f41f1ce9f716b155fc6d59098d21e416ecfeb /var/folders/9f/0snyvwy57519np346scs_fh800fsjh/T/go-build271032812/command-line-arguments.a   

[2018-02-01T21:14:15.440Z] I santad: action=EXEC|decision=ALLOW|reason=PENDING_TRANSITIVE|sha256=pending|pid=32313|ppid=32306|uid=484912|user=nguyenphillip|gid=5000|grou\
p=eng|mode=L|path=/private/var/folders/9f/0snyvwy57519np346scs_fh800fsjh/T/go-build271032812/command-line-arguments/_obj/exe/hello|args=/var/folders/9f/0snyvwy57519np346\
scs_fh800fsjh/T/go-build271032812/command-line-arguments/_obj/exe/hello                                                                                                   

[2018-02-01T21:14:15.452Z] I santad: action=WHITELIST|pid=32312|path=/private/var/folders/9f/0snyvwy57519np346scs_fh800fsjh/T/go-build271032812/command-line-arguments/_o\
bj/exe/hello|sha256=03157283163866e3d7c480ab540bea175551f2e980cde2d4bddd0efe4a4a18a4

nguyen-phillip commented:
PTAL.  Started monitoring file rename events as well as file close because under Xcode 9, if the output file already exists, ld writes to a new file and renames it, but never explicitly closes it.

nguyen-phillip commented:
Adding in some logs to show the behavior of *ld* with Xcode 9. This is every fileop and vnode scope event that occurs during execution of "clang -o hello hello.c"  

1. First time running, no pre-existing output file:

KAUTH_FILEOP_EXEC, uid=484912, pid=22122, ppid=674, process=clang, vnode=0x7fffff806e434c98, path=/usr/bin/clang                                                                                                                                                                                                                                                          
KAUTH_FILEOP_EXEC, uid=484912, pid=22122, ppid=674, process=clang, vnode=0x7fffff8059bb33e0, path=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang                                                                                                                                                                            
KAUTH_VNODE_EXECUTE|, uid=484912, pid=22122, ppid=674, process=clang, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                       
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=22122, ppid=674, process=clang, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                     
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=22122, ppid=674, process=clang, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                     
KAUTH_FILEOP_EXEC, uid=484912, pid=22122, ppid=674, process=clang, vnode=0x7fffff8059bb33e0, path=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang                                                                                                                                                                            
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=22123, ppid=22122, process=clang, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                   
KAUTH_VNODE_READ_DATA|, uid=484912, pid=22123, ppid=22122, process=clang, vp=0x7fffff8053f4b268, dvp=0, path=/KauthTest/hello.c                                                                                                                                                                                                                                           
KAUTH_FILEOP_OPEN, uid=484912, pid=22123, ppid=22122, process=clang, vnode=0x7fffff8053f4b268, path=/KauthTest/hello.c                                                                                                                                                                                                                                                    
KAUTH_FILEOP_CLOSE, uid=484912, pid=22123, ppid=22122, process=clang, vnode=0x7fffff8053f4b268, path=/KauthTest/hello.c, modified=false                                                                                                                                                                                                                                   
KAUTH_FILEOP_EXEC, uid=484912, pid=22122, ppid=674, process=clang, vnode=0x7fffff805abf35d0, path=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld                                                                                                                                                                               
KAUTH_VNODE_WRITE_DATA|, uid=484912, pid=22124, ppid=22122, process=ld, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                     
KAUTH_FILEOP_OPEN, uid=484912, pid=22124, ppid=22122, process=ld, vnode=0x7fffff806f29cba0, path=/KauthTest/hello                                                                                                                                                                                                                                                         
KAUTH_FILEOP_CLOSE, uid=484912, pid=22124, ppid=22122, process=ld, vnode=0x7fffff806f29cba0, path=/KauthTest/hello, modified=true                                                                                                                                                                                                                                         
KAUTH_VNODE_EXECUTE|, uid=484912, pid=22124, ppid=22122, process=ld, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                        
KAUTH_VNODE_WRITE_DATA|, uid=484912, pid=22124, ppid=22122, process=ld, vp=0x7fffff806f29cba0, dvp=0, path=/KauthTest/hello                                                                                                                                                                                                                                               
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=674, ppid=671, process=bash, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                        
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_READ_ATTRIBUTES|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                  
KAUTH_VNODE_READ_ATTRIBUTES|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                  
KAUTH_VNODE_WRITE_DATA|KAUTH_VNODE_APPEND_DATA|KAUTH_VNODE_DELETE_CHILD|KAUTH_VNODE_ACCESS|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                   
KAUTH_VNODE_READ_DATA|KAUTH_VNODE_ACCESS|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                     
KAUTH_VNODE_EXECUTE|KAUTH_VNODE_ACCESS|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                       
KAUTH_VNODE_READ_ATTRIBUTES|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                  
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_EXECUTE|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                        
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff806f29cba0, dvp=0, path=/KauthTest/hello 

2. Second time running, "hello" output file already exists:

KAUTH_FILEOP_EXEC, uid=484912, pid=24788, ppid=674, process=clang, vnode=0x7fffff806e434c98, path=/usr/bin/clang                                                                                                                                                                                                                                                          
KAUTH_FILEOP_EXEC, uid=484912, pid=24788, ppid=674, process=clang, vnode=0x7fffff8059bb33e0, path=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang                                                                                                                                                                            
KAUTH_VNODE_EXECUTE|, uid=484912, pid=24788, ppid=674, process=clang, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                       
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=24788, ppid=674, process=clang, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                     
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=24788, ppid=674, process=clang, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                     
KAUTH_FILEOP_EXEC, uid=484912, pid=24788, ppid=674, process=clang, vnode=0x7fffff8059bb33e0, path=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang                                                                                                                                                                            
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=24789, ppid=24788, process=clang, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                   
KAUTH_VNODE_READ_DATA|, uid=484912, pid=24789, ppid=24788, process=clang, vp=0x7fffff8053f4b268, dvp=0, path=/KauthTest/hello.c                                                                                                                                                                                                                                           
KAUTH_FILEOP_OPEN, uid=484912, pid=24789, ppid=24788, process=clang, vnode=0x7fffff8053f4b268, path=/KauthTest/hello.c                                                                                                                                                                                                                                                    
KAUTH_FILEOP_CLOSE, uid=484912, pid=24789, ppid=24788, process=clang, vnode=0x7fffff8053f4b268, path=/KauthTest/hello.c, modified=false                                                                                                                                                                                                                                   
KAUTH_FILEOP_EXEC, uid=484912, pid=24788, ppid=674, process=clang, vnode=0x7fffff805abf35d0, path=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld                                                                                                                                                                               
KAUTH_VNODE_WRITE_DATA|KAUTH_VNODE_ACCESS|, uid=484912, pid=24790, ppid=24788, process=ld, vp=0x7fffff806f29cba0, dvp=0, path=/KauthTest/hello                                                                                                                                                                                                                            
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=24790, ppid=24788, process=ld, vp=0x7fffff806f29cba0, dvp=0, path=/KauthTest/hello                                                                                                                                                                                                                
KAUTH_VNODE_DELETE|, uid=484912, pid=24790, ppid=24788, process=ld, vp=0x7fffff806f29cba0, dvp=0x7fffff8056e5cba0, path=/KauthTest/hello                                                                                                                                                                                                                                  
KAUTH_FILEOP_DELETE, uid=484912, pid=24790, ppid=24788, process=ld, vnode=0x7fffff806f29cba0, path=/KauthTest/hello                                                                                                                                                                                                                                                       
KAUTH_VNODE_EXECUTE|, uid=484912, pid=24790, ppid=24788, process=ld, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                        
KAUTH_VNODE_WRITE_DATA|, uid=484912, pid=24790, ppid=24788, process=ld, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                     
KAUTH_FILEOP_OPEN, uid=484912, pid=24790, ppid=24788, process=ld, vnode=0x7fffff806f29cba0, path=/KauthTest/hello.ld_mLgH2P                                                                                                                                                                                                                                               
KAUTH_VNODE_CHECKIMMUTABLE|, uid=484912, pid=24790, ppid=24788, process=ld, vp=0x7fffff806f29cba0, dvp=0, path=/KauthTest/hello.ld_mLgH2P                                                                                                                                                                                                                                 
KAUTH_VNODE_WRITE_SECURITY|, uid=484912, pid=24790, ppid=24788, process=ld, vp=0x7fffff806f29cba0, dvp=0, path=/KauthTest/hello.ld_mLgH2P                                                                                                                                                                                                                                 
KAUTH_VNODE_DELETE|, uid=484912, pid=24790, ppid=24788, process=ld, vp=0x7fffff806f29cba0, dvp=0x7fffff8056e5cba0, path=/KauthTest/hello.ld_mLgH2P                                                                                                                                                                                                                        
KAUTH_VNODE_WRITE_DATA|, uid=484912, pid=24790, ppid=24788, process=ld, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                     
KAUTH_FILEOP_RENAME, uid=484912, pid=24790, ppid=24788, process=ld, oldpath=/KauthTest/hello.ld_mLgH2P, newpath=/KauthTest/hello                                                                                                                                                                                                                                          
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=674, ppid=671, process=bash, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                        
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_READ_ATTRIBUTES|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                  
KAUTH_VNODE_READ_ATTRIBUTES|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                  
KAUTH_VNODE_WRITE_DATA|KAUTH_VNODE_APPEND_DATA|KAUTH_VNODE_DELETE_CHILD|KAUTH_VNODE_ACCESS|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                   
KAUTH_VNODE_READ_DATA|KAUTH_VNODE_ACCESS|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                     
KAUTH_VNODE_EXECUTE|KAUTH_VNODE_ACCESS|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                       
KAUTH_VNODE_READ_ATTRIBUTES|, uid=484912, pid=626, ppid=1, process=Finder, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                  
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                      
KAUTH_VNODE_EXECUTE|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                                        
KAUTH_VNODE_READ_DATA|KAUTH_VNODE_ACCESS|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff8056e5cba0, dvp=0, path=/KauthTest                                                                                                                                                                                                                                   
KAUTH_VNODE_READ_ATTRIBUTES|KAUTH_VNODE_READ_SECURITY|, uid=484912, pid=60, ppid=1, process=fseventsd, vp=0x7fffff806f29cba0, dvp=0, path=/KauthTest/hello

tburgin commented:
I would be curious to see what OpenBSM logs look like when performing `"clang -o hello hello.c`

nguyen-phillip commented:
From OpenBSM, looking at all events, but filtering on the path:

1. First time running "clang -o hello hello.c" without existing output file.
 
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang|header, 130, 11, access(2), 0, Fri Jun  1 18:37:26 2018,  + 839 msec|path, hello.c|path, /KauthTest/hello.c|return, success, 0|                                                                                 
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang|header, 144, 11, open(2) - read, 0, Fri Jun  1 18:37:26 2018,  + 847 msec|argument, 2, 0x1000000, flags|path, hello.c|path, /KauthTest/hello.c|return, success, 3|                                              
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang|header, 142, 11, fcntl(2), 0, Fri Jun  1 18:37:26 2018,  + 847 msec|argument, 2, 0x15e, cmd|path, /KauthTest/hello.c|argument, 1, 0x3, fd|return, success, 0|                                                   
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang|header, 130, 11, close(2), 0, Fri Jun  1 18:37:26 2018,  + 848 msec|argument, 2, 0x3, fd|path, /KauthTest/hello.c|return, success, 0|                                                                           
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld|header, 104, 11, open(2) - read,write,creat, 0, Fri Jun  1 18:37:26 2018,  + 883 msec|argument, 3, 0x1ed, mode|argument, 2, 0x202, flags|path, hello|return, success, 3|                                           
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld|header, 128, 11, close(2), 0, Fri Jun  1 18:37:26 2018,  + 884 msec|argument, 2, 0x3, fd|path, /KauthTest/hello|return, success, 0|                                                                                
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld|header, 126, 11, truncate(2), 0, Fri Jun  1 18:37:26 2018,  + 884 msec|path, hello|path, /KauthTest/hello|return, success, 0|                                                                                      
process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:37:27 2018,  + 347 msec|path, /KauthTest|path, /KauthTest|return, success, 0|                                                                                                                 
process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:37:27 2018,  + 347 msec|path, /KauthTest|path, /KauthTest|return, success, 0|                                                                                                                 
process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:37:27 2018,  + 348 msec|path, /KauthTest|path, /KauthTest|return, success, 0|                                                                                                           
                           
2. Second run of "clang -o hello hello.c" executed immediately after.
                                                                                                                                                                                                                                                                                          
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang|header, 130, 11, access(2), 0, Fri Jun  1 18:37:33 2018,  + 260 msec|path, hello.c|path, /KauthTest/hello.c|return, success, 0|                                                                                 
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang|header, 144, 11, open(2) - read, 0, Fri Jun  1 18:37:33 2018,  + 267 msec|argument, 2, 0x1000000, flags|path, hello.c|path, /KauthTest/hello.c|return, success, 3|                                              
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang|header, 142, 11, fcntl(2), 0, Fri Jun  1 18:37:33 2018,  + 267 msec|argument, 2, 0x15e, cmd|path, /KauthTest/hello.c|argument, 1, 0x3, fd|return, success, 0|                                                   
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang|header, 130, 11, close(2), 0, Fri Jun  1 18:37:33 2018,  + 267 msec|argument, 2, 0x3, fd|path, /KauthTest/hello.c|return, success, 0|                                                                           
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld|header, 126, 11, access(2), 0, Fri Jun  1 18:37:33 2018,  + 300 msec|path, hello|path, /KauthTest/hello|return, success, 0|                                                                                        
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld|header, 126, 11, access(2), 0, Fri Jun  1 18:37:33 2018,  + 300 msec|path, hello|path, /KauthTest/hello|return, success, 0|                                                                                        
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld|header, 126, 11, unlink(2), 0, Fri Jun  1 18:37:33 2018,  + 300 msec|path, hello|path, /KauthTest/hello|return, success, 0|                                                                                        
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld|header, 114, 11, openat(2) - read,write,create, 0, Fri Jun  1 18:37:33 2018,  + 301 msec|argument, 4, 0x180, mode|argument, 3, 0xa02, flags|path, hello.ld_JHAOc5|return, success, 3|                              
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld|header, 138, 11, ftruncate(2), 0, Fri Jun  1 18:37:33 2018,  + 301 msec|path, /KauthTest/hello.ld_JHAOc5|argument, 1, 0x3, fd|return, success, 0|                                                                  
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld|header, 168, 11, chmod(2), 0, Fri Jun  1 18:37:33 2018,  + 301 msec|argument, 2, 0x1ed, new file mode|path, hello.ld_JHAOc5|path, /KauthTest/hello.ld_JHAOc5|return, success, 0|                                   
process=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld|header, 155, 11, rename(2), 0, Fri Jun  1 18:37:33 2018,  + 301 msec|path, hello.ld_JHAOc5|path, /KauthTest/hello.ld_JHAOc5|path, hello|return, success, 0|                                                        
process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:37:33 2018,  + 736 msec|path, /KauthTest|path, /KauthTest|return, success, 0|                                                                                                                 
process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:37:33 2018,  + 736 msec|path, /KauthTest|path, /KauthTest|return, success, 0|                                                                                                                 
process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:37:33 2018,  + 736 msec|path, /KauthTest|path, /KauthTest|return, success, 0|                                                                                                                 
process=/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/FSEvents.framework/Versions/A/Support/fseventsd|header, 125, 11, access(2), 0, Fri Jun  1 18:37:34 2018,  + 236 msec|path, /KauthTest|path, /KauthTest|return, success, 0|       

3. And a separate test, this is opening a file (hello_stream.txt), writing to it incrementally over a period of a few seconds, then exiting without an explicit close().

process=/usr/libexec/taskgated|header, 139, 11, open(2) - read, 0, Fri Jun  1 18:40:18 2018,  + 303 msec|argument, 2, 0x1100004, flags|path, /KauthTest|path, /KauthTest|return, success, 3|
process=/usr/libexec/taskgated|header, 122, 11, close(2), 0, Fri Jun  1 18:40:18 2018,  + 303 msec|argument, 2, 0x3, fd|path, /KauthTest|return, success, 0|
process=/usr/libexec/taskgated|header, 139, 11, open(2) - read, 0, Fri Jun  1 18:40:18 2018,  + 303 msec|argument, 2, 0x1100004, flags|path, /KauthTest|path, /KauthTest|return, success, 3|
process=/usr/libexec/taskgated|header, 122, 11, close(2), 0, Fri Jun  1 18:40:18 2018,  + 303 msec|argument, 2, 0x3, fd|path, /KauthTest|return, success, 0|
process=/usr/libexec/taskgated|header, 153, 11, open(2) - read, 0, Fri Jun  1 18:40:18 2018,  + 303 msec|argument, 2, 0x0, flags|path, /KauthTest/stream|path, /KauthTest/stream|return, success, 3|
process=/usr/libexec/taskgated|header, 129, 11, close(2), 0, Fri Jun  1 18:40:18 2018,  + 305 msec|argument, 2, 0x3, fd|path, /KauthTest/stream|return, success, 0|
process=/KauthTest/stream|header, 144, 11, execve(2), 0, Fri Jun  1 18:40:18 2018,  + 305 msec|exec arg, ./stream|path, ./stream|path, /KauthTest/stream|return, success, 0|
process=/KauthTest/stream|header, 130, 11, open(2) - read, 0, Fri Jun  1 18:40:18 2018,  + 306 msec|argument, 2, 0x0, flags|path, .|path, /KauthTest|return, success, 3|
process=/KauthTest/stream|header, 134, 11, fcntl(2), 0, Fri Jun  1 18:40:18 2018,  + 306 msec|argument, 2, 0x15e, cmd|path, /KauthTest|argument, 1, 0x3, fd|return, success, 0|
process=/KauthTest/stream|header, 122, 11, close(2), 0, Fri Jun  1 18:40:18 2018,  + 306 msec|argument, 2, 0x3, fd|path, /KauthTest|return, success, 0|
process=/KauthTest/stream|header, 115, 11, open(2) - write,creat,trunc, 0, Fri Jun  1 18:40:18 2018,  + 308 msec|argument, 3, 0x1a4, mode|argument, 2, 0x601, flags|path, hello_stream.txt|return, success, 3|
process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:40:18 2018,  + 450 msec|path, /KauthTest|path, /KauthTest|return, success, 0|
process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:40:18 2018,  + 450 msec|path, /KauthTest|path, /KauthTest|return, success, 0|
process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:40:18 2018,  + 450 msec|path, /KauthTest|path, /KauthTest|return, success, 0|

process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:40:29 2018,  + 625 msec|path, /KauthTest|path, /KauthTest|return, success, 0|
process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:40:29 2018,  + 625 msec|path, /KauthTest|path, /KauthTest|return, success, 0|
process=/System/Library/CoreServices/Finder.app/Contents/MacOS/Finder|header, 125, 11, getattrlist(), 0, Fri Jun  1 18:40:29 2018,  + 626 msec|path, /KauthTest|path, /KauthTest|return, success, 0|

russellhancox commented on Tests/KernelTests/main.mm:
```diff
@@ -57,6 +57,9 @@
       exit(1); \
     } while (0)
 
+const char *ldPath =
```
This is brittle. Instead of hardcoding, use this:

```xcrun -f ld```, which prints the full path to the _active_ ld

nguyen-phillip replied:
Done.

russellhancox commented on Tests/KernelTests/main.mm:
```diff
@@ -107,20 +107,26 @@ - (NSString *)sha256ForPath:(NSString *)path {
+      IOConnectCallStructMethod(self.connection, kSantaUserClientAllowBinary,
+                                &vnodeid, sizeof(vnodeid), 0, 0);
+      break;
+  case ACTION_RESPOND_DENY:
```
case indents are wrong

nguyen-phillip replied:
Done.
